### PR TITLE
Installation step for iroha

### PR DESCRIPTION
**Pull request by Warchant**
This PR adds `package` step for iroha. On debian-based OS it will create `iroha-0.95.0-Linux.{tar.gz,deb}` packages.

Deb package has its own internal dependencies (like `libboost-system-dev`, etc) which should be installed manually using `apt`.

However, `libgrpc`, `libprotobuf` and `libgrpc++` can not be installed through apt, and I decided to package them together with binaries.

Currently, anything can be packaged (directory, file, library, binary, whatever) just by adding according `install` step, thanks to CPack.

Changelog:
- [x] tar.gz package
- [x] deb package
- [ ] MAC package (current version has some bundle code, but it is not tested by myself)
- [x] rpm bundle (untested)
- [x] ed25519 is now static lib
- [x] if some dependency is not installed, cmake now pretty prints what exactly is required with link to github and commit id.
- [x] added RPATH to this folders: `.` , `lib`, `../lib`. It means, that libraries can be in any of these directories.

Example:
```
# it will work only in Release now. Should it be changed?
$ cmake .. -DCMAKE_BUILD_TYPE=Release
$ make -j8 package

[100%] Built target tx_pipeline_integration_test
Run CPack packaging tool...
CPack: Create package using TGZ
CPack: Install projects
CPack: - Run preinstall target for: iroha
CPack: - Install project: iroha
CPack: Create package
CPack: - package: /home/bogdan/tools/iroha/build/iroha-0.95.0-Linux.tar.gz generated.
CPack: Create package using DEB
CPack: Install projects
CPack: - Run preinstall target for: iroha
CPack: - Install project: iroha
CPack: Create package
CPack: - package: /home/bogdan/tools/iroha/build/iroha-0.95.0-Linux.deb generated.

$ tar -xf /home/bogdan/tools/iroha/build/iroha-0.95.0-Linux.tar.gz
$ cd iroha-0.95.0-Linux
$ tree .
bogdan@xps:~/tools/iroha/build/iroha-0.95.0-Linux$ tree .
.
├── bin
│   ├── iroha-cli
│   └── irohad
└── lib
    ├── libgrpc++_reflection.so.1
    ├── libgrpc++.so.1
    ├── libgrpc.so.3
    └── libprotobuf.so.13
```

### CODE REVIEW

**Comment by neewy** [comment]
You may forget about mac package, I can create it via brew, or we can distribute a tar.gz with binaries compiled in macOS environment

File: cmake/Modules/Findrapidjson.cmake [hunk 1]
```diff
@@ -3,15 +3,18 @@ add_library(rapidjson INTERFACE IMPORTED)
 find_path(rapidjson_INCLUDE_DIR rapidjson/rapidjson.h)
 mark_as_advanced(rapidjson_INCLUDE_DIR)
 
-find_package(PackageHandleStandardArgs REQUIRED)
 find_package_handle_standard_args(rapidjson DEFAULT_MSG
     rapidjson_INCLUDE_DIR
     )
 
+set(GITHUB https://github.com/miloyip/rapidjson.git)
```
Comment (l4l) [inline cmake/Modules/Findrapidjson.cmake]: u mean URL, i guess?
check the following as well plz

File: cmake/Modules/Findcpp_redis.cmake [hunk 2]
```diff
@@ -9,17 +9,20 @@ mark_as_advanced(cpp_redis_LIBRARY)
...
 mark_as_advanced(cpp_redis_LIBRARY)
 
-find_package(PackageHandleStandardArgs REQUIRED)
 find_package_handle_standard_args(cpp_redis DEFAULT_MSG
     cpp_redis_INCLUDE_DIR
     cpp_redis_LIBRARY
     tacopie_LIBRARY
     )
 
+set(URL https://github.com/Cylix/tacopie.git)
```
Comment (l4l) [inline cmake/Modules/Findcpp_redis.cmake]: url changed, comments plz

File: CMakeLists.txt [hunk 3]
```diff
@@ -8,10 +8,13 @@ endif()
 
 PROJECT(iroha C CXX)
 
-SET(CMAKE_CXX_FLAGS "-std=c++1y -Wall -fPIC")
+SET(CMAKE_CXX_FLAGS "-std=c++1y -Wall -fPIC -fPIE")
```
Comment (l4l) [inline CMakeLists.txt]: not a good idea, ping @lebdron 
afair some compilers use -Wl,-pie or -pie

File: cmake/functions.cmake [hunk 4]
```diff
@@ -65,3 +66,64 @@ function(compile_proto_to_grpc_cpp PROTO)
...
+      COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
+      OUTPUT_VARIABLE ${commit}
+      WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+  )
+endmacro()
+
+
+macro(get_current_architecture arch)
+  execute_process(
+      COMMAND dpkg --print-architecture
```
Comment (l4l) [inline cmake/functions.cmake]: dpkg? orly?
use `uname -m` or specify that is debian specific

File: cmake/release.cmake [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Iroha - simple decentralized blockchain")
+
+SET(CPACK_PACKAGE_VENDOR              "Soramitsu LLC")
+SET(CPACK_RESOURCE_FILE_LICENSE       "${CMAKE_SOURCE_DIR}/LICENSE")
+SET(CPACK_PACKAGE_CONTACT             "Bogdan Vaneev <bogdan@soramitsu.co.jp>")
+
+# change build type
+SET(CMAKE_BUILD_TYPE Release)
+
+if(NOT IROHA_VERSION)
```
Comment (l4l) [inline cmake/release.cmake]: `IROHA_VERSION` goes in #610 don't mix this, just use hash for now

File: cmake/release.cmake [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+SET(CPACK_PACKAGE_VERSION ${IROHA_VERSION})
+message(STATUS "[IROHA_VERSION] '${IROHA_VERSION}'")
+
+
+SET(CPACK_STRIP_FILES TRUE)
+
+set(CPACK_COMPONENTS_ALL binaries libraries)
+
+if(WIN32)
```
Comment (l4l) [inline cmake/release.cmake]: we don't plan to support it
remove it plz, anyway it's empty

File: cmake/release/linux-rpm.cmake [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+else()
+  # add deb to list of generators
+  list(APPEND CPACK_GENERATOR RPM)
+
+  get_current_architecture(arch)
+
+  message(STATUS "[package-${arch}] linux-rpm standalone will be packaged. Use -DNO_RPM to disable.")
+
+  SET(CPACK_RPM_PACKAGE_LICENSE "${CMAKE_SOURCE_DIR}/LICENSE")
+  SET(CPACK_RPM_PACKAGE_REQUIRES "libstdc++5, libpq-dev, libtbb-dev, libboost-system-dev, libboost-filesystem-dev, libc-ares-dev")
```
Comment (l4l) [inline cmake/release/linux-rpm.cmake]: it works fine with libstdc++6

**Comment by neewy** [comment]
I confirm that this creates valid binaries for macOS

File: cmake/functions.cmake [hunk 8]
```diff
@@ -65,3 +66,60 @@ function(compile_proto_to_grpc_cpp PROTO)
...
+
+macro(add_install_step_for_bin target)
+  install(TARGETS ${target}
+      RUNTIME DESTINATION bin
+      CONFIGURATIONS Release
+      COMPONENT binaries)
+endmacro()
+
+
+# does not work (yet) for system libraries, since we need root
```
Comment (neewy) [inline cmake/functions.cmake]: What is that?

File: cmake/release.cmake [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+
+SET(CPACK_STRIP_FILES TRUE)
+
+set(CPACK_COMPONENTS_ALL binaries libraries)
+
+if (APPLE)
+  message(WARNING "On OSX only TGZ/ZIP packaging is supported")
+  # cmake is running on mac os
+#  include(cmake/release/osx-bundle.cmake)
```
Comment (neewy) [inline cmake/release.cmake]: what is that

File: cmake/release/linux/deb/iroha.cmake [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  message(STATUS "[package-${arch}] linux-deb standalone will be packaged.")
+
+  set(CPACK_DEBIAN_PACKAGE_NAME          iroha  )
+  set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE  ${arch})  # dpkg --print-architecture
+  set(CPACK_DEBIAN_PACKAGE_DEPENDS "libtbb-dev")
+  set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS       ON)
+  set(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS ON)
+  set(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS_POLICY ON)
+#  set(CPACK_DEB_COMPONENT_INSTALL ON)
```
Comment (neewy) [inline cmake/release/linux/deb/iroha.cmake]: what is that

File: cmake/release/linux/rpm/iroha.cmake [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  # add deb to list of generators
+  list(APPEND CPACK_GENERATOR RPM)
+
+  message(STATUS "[package] linux-rpm standalone will be packaged.")
+
+  SET(CPACK_RPM_PACKAGE_LICENSE "${CMAKE_SOURCE_DIR}/LICENSE")
+  SET(CPACK_RPM_PACKAGE_REQUIRES "libstdc++5, libpq-dev, libtbb-dev, libboost-system-dev, libboost-filesystem-dev, libc-ares-dev")
+endif()
+
+#   SET(CPACK_RPM_PACKAGE_REQUIRES "libstdc++5, libpq-dev (>= 9.5), libtbb-dev (>= 4.4), libboost-system-dev (>= 1.56.0), libboost-filesystem-dev (>= 1.56.0), libc-ares-dev (>= 1.10.0)")
```
Comment (neewy) [inline cmake/release/linux/rpm/iroha.cmake]: what is that

File: cmake/release/osx/osx-bundle.cmake [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#message(STATUS "standalone osx-bundle can be packaged")
+#
+## TODO (@warchant): not tested. should be tested on mac
```
Comment (neewy) [inline cmake/release/osx/osx-bundle.cmake]: what is that? tar.gz is enough

File: scripts/swarm-deploy.sh [hunk 13]
```diff
@@ -69,9 +72,9 @@ do
     # copy config
     docker cp ${IROHA_HOME}/iroha.conf ${COMPOSE_PROJECT_NAME}_node_1:/
     # generate genesis block based on list of peers
-    docker exec ${COMPOSE_PROJECT_NAME}_node_1 /opt/iroha/iroha-cli --genesis_block --peers_address /peers.list
+    docker exec ${COMPOSE_PROJECT_NAME}_node_1 $cli --genesis_block --peers_address /peers.list
```
Comment (neewy) [inline scripts/swarm-deploy.sh]: did you do that with @lebdron? or just simply rebased?

Comment (Warchant) [reply scripts/swarm-deploy.sh]: I changed it.

File: cmake/release/osx/osx-bundle.cmake (hunk 12 above)
Comment (Warchant) [reply cmake/release/osx/osx-bundle.cmake]: Okay. It can be removed.

File: cmake/release/linux/rpm/iroha.cmake (hunk 11 above)
Comment (Warchant) [reply cmake/release/linux/rpm/iroha.cmake]: It is there to not forget required versions. 
Actually, can be safely removed.

File: cmake/release/linux/deb/iroha.cmake (hunk 10 above)
Comment (Warchant) [reply cmake/release/linux/deb/iroha.cmake]: Can be safely removed

File: cmake/release.cmake (hunk 9 above)
Comment (Warchant) [reply cmake/release.cmake]: Can be safely removed.

File: cmake/functions.cmake (hunk 8 above)
Comment (Warchant) [reply cmake/functions.cmake]: That macro can be removed.

File: cmake/release.cmake (hunk 6 above)
Comment (neewy) [reply cmake/release.cmake]: I agree

**Comment by l4l** [comment]
Overall code looks good, though didn't tested creation of packages

**Pull request merged**